src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java

```diff
@@ -229,29 +243,82 @@ public boolean isThrottleMatrixConfigurations() {
+
+    static List<ThrottledStepInfo> getThrottledPipelinesForCategory(String category) {
+        List<ThrottledStepInfo> throttledPipelines = new ArrayList<>();
+
+        DescriptorImpl descriptor = Jenkins.getActiveInstance().getDescriptorByType(DescriptorImpl.class);
+        for (Map.Entry<String,Integer> currentPipeline : descriptor.getThrottledPipelinesForCategory(category).entrySet()) {
+            Run<?,?> flowNodeRun = Run.fromExternalizableId(currentPipeline.getKey());
+
+            if (flowNodeRun == null) {
+                // No run found, so remove the throttle.
+                descriptor.removeThrottledPipelineForCategory(currentPipeline.getKey(), category, null);
+            } else if (!(flowNodeRun instanceof FlowExecutionOwner.Executable)) {
+                // If for some reason we've somehow ended up with a non-pipeline job, remove the throttle.
+                descriptor.removeThrottledPipelineForCategory(currentPipeline.getKey(), category, null);
+            } else if (!flowNodeRun.isBuilding()) {
+                // The run is done building, so remove the throttle.
+                descriptor.removeThrottledPipelineForCategory(currentPipeline.getKey(), category, null);
+            } else {
+                FlowExecutionOwner owner = ((FlowExecutionOwner.Executable)flowNodeRun).asFlowExecutionOwner();
+                FlowExecution execution = owner.getOrNull();
+                if (execution == null) {
+                    // For some reason, the flow execution is null, so again? Remove the throttle.
+                    descriptor.removeThrottledPipelineForCategory(currentPipeline.getKey(), category, null);
+                } else {
+                    try {
+                        for (StepExecution stepExec : execution.getCurrentExecutions(false).get()) {
+                            try {
+                                Computer c = stepExec.getContext().get(Computer.class);
+                                ThrottledStepInfo candidateInfo = stepExec.getContext().get(ThrottledStepInfo.class);
+                                if (c != null && candidateInfo != null) {
+                                    ThrottledStepInfo info = candidateInfo.forCategory(category);
+                                    if (info != null) {
+                                        if (info.getNode() == null && c.getNode() != null) {
+                                            info.setNode(c.getNode().getNodeName());
+                                        }
+                                        throttledPipelines.add(info);
+                                    }
+                                }
+                            } catch (IOException e) {
+                                // TODO: What do we do here if anything?
+                            }
+                        }
+                    } catch (InterruptedException | ExecutionException e) {
+                        // TODO: What do we do here if anything?
```

Comment:
If you are calling `Future.get` that is a bad sign. Means you are causing this thread, whatever it is, to block on the CPS VM thread, and potentially to even wait indefinitely for a build to rehydrate its pickles.

Use a callback on the future instead, which means refactoring everything else to be asynchronous.